### PR TITLE
Set - 7: UUID

### PR DESCRIPTION
**Pull request by edualonso**
- Added support for UUID.
- Improved type checks for foreign types.
- Added missing test cases for collections of foreign types.
- Added missing logic for Set interface contract regarding type-checking collection types.

### CODE REVIEW

File: realm/realm-library/src/main/java/io/realm/SetValueOperator.java [hunk 1]
```diff
@@ -533,6 +536,82 @@ boolean removeAllInternal(Collection<?> c) {
...
+        return osSet.add(value);
+    }
+
+    @Override
+    boolean contains(@Nullable Object o) {
+        UUID value;
+        if (o == null) {
+            value = null;
+        } else {
+            value = (UUID) o;
```
Comment (rorbech) [inline realm/realm-library/src/main/java/io/realm/SetValueOperator.java:559]: Is this cast safe. I though I checked up on it in one of the other PRs, but smells like you could actually get called with arbitrary types all the way through to here. 

Similar on the other operations.

File: realm/realm-library/src/main/java/io/realm/SetValueOperator.java [hunk 2]
```diff
@@ -533,6 +536,82 @@ boolean removeAllInternal(Collection<?> c) {
...
+            }
+        }
+
+        NativeMixedCollection collection = NativeMixedCollection.newUUIDCollection(c);
+        return osSet.collectionFunnel(collection, OsSet.ExternalCollectionOperation.ADD_ALL);
+    }
+
+    @Override
+    boolean removeAllInternal(Collection<?> c) {
+        //noinspection unchecked
```
Comment (rorbech) [inline realm/realm-library/src/main/java/io/realm/SetValueOperator.java:602]: We are also not guaranteed to have an UUID collection here, aren't we?

Similarly for all the other collection methods.

**Review by rorbech — COMMENTED** [review]
There some concerning casts. Maybe I am just missing the overview (I though I checked them as part of the previous PR, maybe I am missing the bigger picture in this PR). If it is not guarded by the type system that we can only call `contains`, `containsAll`, etc. we should definitely have tests for raising appropriate errors.

File: realm/realm-library/src/main/java/io/realm/SetValueOperator.java (hunk 1 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/SetValueOperator.java:559]: Fixed, also other operations.

File: realm/realm-library/src/main/java/io/realm/SetValueOperator.java (hunk 2 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/SetValueOperator.java:671]: Fixed.

**Comment by edualonso** [comment]
I've corrected the type casting issues and improved a few other things, including missing tests.

File: realm/realm-library/src/androidTest/kotlin/io/realm/ManagedSetTester.kt [hunk 3]
```diff
@@ -102,8 +102,13 @@ class ManagedSetTester<T : Any>(
         }
         assertFalse(set.contains(notPresentValue))
 
-        // Check contains with something entirely different
-        assertFalse(set.contains(Pair(1, 2) as T))
+        // Throws if we call contains with something entirely different
+        val somethingEntirelyDifferent = initializedSet.map {
+            Pair(it, it)
+        }
+        assertFailsWith<ClassCastException> {
```
Comment (rorbech) [inline realm/realm-library/src/androidTest/kotlin/io/realm/ManagedSetTester.kt:109]: 🎉

**Review by rorbech — APPROVED** [review]
Looks like the test that `add(incompatibleType)` will throw is stil missing 👀 

Could be that we could have relied on the JVM's ClassCastException and skip some checks/iterations, but I like that we can write our own exceptions.

**Comment by edualonso** [comment]
I'm not sure we can call add with an incompatible type. The function signature is `boolean add(E e)` so it's pretty tied to the type we declare in the set itself.

**Comment by rorbech** [comment]
> The function signature is `boolean add(E e)` so it's pretty tied to the type we declare in the set itself.

Ahh. That was probably where my confusion arose. I must have somehow mixed the "non-throwing" bounded method returning a boolean, with the throwing unbounded ones when I reviewed. Again, sorry for the confusion.

**Pull request merged**